Unavailable packages are silently ignored by `opam show`

... when passing more than one package.

If _all_ of the packages passed to `opam` show are non-existent, an error is shown:

```bash
ᐅ opam show --field=version -- void
[ERROR] No package matching void found

ᐅ opam show --field=version -- void1 void2
[ERROR] No package matching void1 or void2 found
```

But if only _some_ packages are non-existent, the failure is ignored:

```bash
ᐅ opam show --field=version -- dune void
2.5.0
```

I would expect error messages to be displayed individually and the exit-code to be non-zero if any error messages have been displayed.

--- a/master_changes.md
+++ b/master_changes.md
@@ -1,5 +1,6 @@
 Working version changelog, used as a base for the changelog and the release
 note.
+Possibly scripts breaking changes are prefixed with ✘
 
 ## Admin
   * Fix admin cache synchronisation message [#4193 @rjbou - fix #4167]
@@ -23,6 +24,12 @@ note.
 
 ## Pin
   * Don't keep unpinned package version if it exists in repo [#4073 @rjbou - fix #3630]
+
+## Show
+  * ✘ Display error message for all not found packages [#4179 @rjbou - fix #4164]
+  * ✘ Keep package order given via cli [#4179 @rjbou - fix #4163]
+  * `--sort`` apply to with all options, not only `--just-file` [#4179 @rjbou]
+
 
 ## Depext
   * Fix performance issue of depext under Docker/debian [#4165 @AltGr]

--- a/src/client/opamCommands.ml
+++ b/src/client/opamCommands.ml
@@ -774,7 +774,7 @@ let show =
         OpamConsole.error_and_exit `Not_found "No package found"
       else
         OpamListCommand.info st
-          ~fields ~raw ~where ~normalise ~show_empty ~all_versions atoms;
+          ~fields ~raw ~where ~normalise ~show_empty ~all_versions ~sort atoms;
       `Ok ()
     | atom_locs, true ->
       if List.exists (function `Atom _ -> true | _ -> false) atom_locs then

--- a/src/client/opamListCommand.ml
+++ b/src/client/opamListCommand.ml
@@ -716,9 +716,8 @@ let info st ~fields ~raw ~where ?normalise ?(show_empty=false)
     OpamFormula.packages_of_atoms ~disj:all_versions
       (st.packages ++ st.installed) atoms
   in
-  let missing_atoms =
-    OpamStd.List.filter_map (fun ((n,_) as atom) ->
-        if OpamPackage.has_name packages n then None else Some atom) atoms
+  let atoms, missing_atoms =
+    List.partition (fun (n,_) -> OpamPackage.has_name packages n) atoms
   in
   if missing_atoms <> [] then
     (OpamConsole.error "No package matching %s found"
@@ -807,8 +806,7 @@ let info st ~fields ~raw ~where ?normalise ?(show_empty=false)
     | [f] -> OpamConsole.msg "%s\n" (detail_printer ?normalise st pkg f)
     | fields -> output_table fields pkg
   in
-  OpamPackage.names_of_packages packages |>
-  OpamPackage.Name.Set.iter (fun name ->
+  List.iter (fun (name,_) ->
       (* Like OpamSwitchState.get_package, but restricted to [packages] *)
       let nvs = OpamPackage.packages_of_name packages name in
       if all_versions then
@@ -824,4 +822,4 @@ let info st ~fields ~raw ~where ?normalise ?(show_empty=false)
          in
          header choose;
          output_package choose)
-    )
+    ) atoms

--- a/src/client/opamListCommand.ml
+++ b/src/client/opamListCommand.ml
@@ -716,10 +716,16 @@ let info st ~fields ~raw ~where ?normalise ?(show_empty=false)
     OpamFormula.packages_of_atoms ~disj:all_versions
       (st.packages ++ st.installed) atoms
   in
-  if OpamPackage.Set.is_empty packages then
+  let missing_atoms =
+    OpamStd.List.filter_map (fun ((n,_) as atom) ->
+        if OpamPackage.has_name packages n then None else Some atom) atoms
+  in
+  if missing_atoms <> [] then
     (OpamConsole.error "No package matching %s found"
-       (OpamStd.List.concat_map " or " OpamFormula.short_string_of_atom atoms);
-     OpamStd.Sys.exit_because `Not_found);
+       (OpamStd.List.concat_map " or " OpamFormula.short_string_of_atom
+          missing_atoms);
+     if OpamPackage.Set.is_empty packages then
+       OpamStd.Sys.exit_because `Not_found);
   let fields = List.map (field_of_string ~raw) fields in
   let all_versions_fields = [
     Name;

--- a/src/client/opamListCommand.ml
+++ b/src/client/opamListCommand.ml
@@ -490,12 +490,17 @@ let mini_field_printer ?(prettify=false) ?(normalise=false) =
   | List (_, l) -> OpamPrinter.value_list l
   | f -> OpamPrinter.Normalise.value f
 
-let detail_printer ?prettify ?normalise st nv =
+let detail_printer ?prettify ?normalise ?(sort=false) st nv =
   let open OpamStd.Option.Op in
   let (%) s cols = OpamConsole.colorise' cols s in
   let root_sty =
     if OpamPackage.Set.mem nv st.installed_roots then [`underline]
     else []
+  in
+  let get_opam =
+    if not sort then get_opam else
+    fun st nv ->
+      OpamFileTools.sort_opam (get_opam st nv)
   in
   function
   | Name -> OpamPackage.Name.to_string nv.name % (`bold :: root_sty)
@@ -711,7 +716,7 @@ let print_depexts =
   OpamSysPkg.Set.iter (fun d -> OpamConsole.msg "%s\n" (OpamSysPkg.to_string d))
 
 let info st ~fields ~raw ~where ?normalise ?(show_empty=false)
-    ?(all_versions=false) atoms =
+    ?(all_versions=false) ?(sort=false) atoms =
   let packages =
     OpamFormula.packages_of_atoms ~disj:all_versions
       (st.packages ++ st.installed) atoms
@@ -756,7 +761,7 @@ let info st ~fields ~raw ~where ?normalise ?(show_empty=false)
   let output_table fields nv =
     let tbl =
       List.fold_left (fun acc item ->
-          let contents = detail_printer ?normalise st nv item in
+          let contents = detail_printer ?normalise ~sort st nv item in
           if show_empty || contents <> "" then
             [ OpamConsole.colorise `blue (string_of_field ~raw item); contents ]
             :: acc
@@ -779,6 +784,10 @@ let info st ~fields ~raw ~where ?normalise ?(show_empty=false)
   in
   let output_package pkg =
     let opam = get_opam st pkg in
+    let opam =
+      if not sort then opam else
+        OpamFileTools.sort_opam opam
+    in
     OpamFile.OPAM.print_errors opam;
     if where then
       OpamConsole.msg "%s\n"
@@ -803,7 +812,7 @@ let info st ~fields ~raw ~where ?normalise ?(show_empty=false)
     | [] ->
       OpamConsole.header_msg "Version-specific details";
       output_table one_version_fields pkg
-    | [f] -> OpamConsole.msg "%s\n" (detail_printer ?normalise st pkg f)
+    | [f] -> OpamConsole.msg "%s\n" (detail_printer ?normalise ~sort st pkg f)
     | fields -> output_table fields pkg
   in
   List.iter (fun (name,_) ->

--- a/src/client/opamListCommand.mli
+++ b/src/client/opamListCommand.mli
@@ -141,7 +141,7 @@ val display: 'a switch_state -> package_listing_format -> package_set -> unit
 val info:
   'a switch_state ->
   fields:string list -> raw:bool -> where:bool ->
-  ?normalise:bool -> ?show_empty:bool -> ?all_versions:bool ->
+  ?normalise:bool -> ?show_empty:bool -> ?all_versions:bool -> ?sort:bool ->
   atom list -> unit
 
 (** Prints the value of an opam field in a shortened way (with [prettify] -- the

